File: frontend/src/api/lookup.ts
import apiClient from './index'

export interface IPLookupRequest {
  ip_address: string
}

export interface IPLookupResult {
  target_ip: string
  found: boolean
  mac_address?: string
  switch_id?: number
  switch_name?: string
  switch_ip?: string
  port_name?: string
  vlan_id?: number
  query_time_ms: number
  message?: string
}

export interface IPLookupResponse {
  success: boolean
  result?: IPLookupResult
  error?: string
}

export interface QueryHistoryItem {
  id: number
  target_ip: string
  found_mac?: string
  switch_id?: number
  port_name?: string
  vlan_id?: number
  query_status: 'success' | 'not_found' | 'error'
  error_message?: string
  query_time_ms?: number
  queried_at: string
}

export interface QueryHistoryListResponse {
  total: number
  page: number
  page_size: number
  items: QueryHistoryItem[]
}

export const lookupApi = {
  // Lookup an IP address
  lookupIP: async (ipAddress: string): Promise<IPLookupResponse> => {
    const response = await apiClient.post('/api/v1/lookup/ip', {
      ip_address: ipAddress
    })
    return response.data
  },

  // Get query history
  getHistory: async (page: number = 1, pageSize: number = 20): Promise<QueryHistoryListResponse> => {
    const response = await apiClient.get('/api/v1/history', {
      params: { page, page_size: pageSize }
    })
    return response.data
  },

  // Get specific history item
  getHistoryItem: async (id: number): Promise<QueryHistoryItem> => {
    const response = await apiClient.get(`/api/v1/history/${id}`)
    return response.data
  }
}

File: backend/src/schemas/lookup.py
from pydantic import BaseModel, Field, IPvAnyAddress
from typing import Optional
from datetime import datetime


class IPLookupRequest(BaseModel):
    """Schema for IP lookup request"""
    ip_address: IPvAnyAddress = Field(..., description="Target IP address to lookup")


class IPLookupResult(BaseModel):
    """Schema for IP lookup result"""
    target_ip: str
    found: bool
    mac_address: Optional[str] = None
    switch_id: Optional[int] = None
    switch_name: Optional[str] = None
    switch_ip: Optional[str] = None
    port_name: Optional[str] = None
    vlan_id: Optional[int] = None
    query_time_ms: int
    query_mode: Optional[str] = None  # 'cache', 'realtime', 'auto_fallback'
    data_age_seconds: Optional[int] = None  # Only for cache mode - how old is the data
    last_seen: Optional[str] = None  # Only for cache mode - when was the data last collected
    message: Optional[str] = None
    freshness: Optional[dict] = None


class IPLookupResponse(BaseModel):
    """Schema for IP lookup response"""
    success: bool
    result: Optional[IPLookupResult] = None
    error: Optional[str] = None
